homeassistant/components/rest/__init__.py

```diff
@@ -1,4 +1,152 @@
+    CONF_AUTHENTICATION,
+    CONF_HEADERS,
+    CONF_METHOD,
+    CONF_PARAMS,
+    CONF_PASSWORD,
+    CONF_PAYLOAD,
+    CONF_RESOURCE,
+    CONF_RESOURCE_TEMPLATE,
+    CONF_SCAN_INTERVAL,
+    CONF_TIMEOUT,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+    HTTP_DIGEST_AUTHENTICATION,
+    SERVICE_RELOAD,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import discovery
+from homeassistant.helpers.entity_component import (
+    DEFAULT_SCAN_INTERVAL,
+    EntityComponent,
+)
+from homeassistant.helpers.reload import async_reload_integration_platforms
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import COORDINATOR, DOMAIN, REST, SHARED_DATA_ID
+from .data import RestData
+from .schema import CONFIG_SCHEMA  # noqa:F401 pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
 PLATFORMS = ["binary_sensor", "notify", "sensor", "switch"]
+COORDINATOR_AWARE_PLATFORMS = [SENSOR_DOMAIN, BINARY_SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the rest platforms."""
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+    hass.data[DOMAIN] = {platform: {} for platform in COORDINATOR_AWARE_PLATFORMS}
+
+    async def reload_service_handler(service):
+        """Remove all user-defined groups and load new ones from config."""
+        conf = await component.async_prepare_reload()
+        if conf is None:
+            return
+        await async_reload_integration_platforms(hass, DOMAIN, PLATFORMS)
+        hass.data[DOMAIN] = {platform: {} for platform in COORDINATOR_AWARE_PLATFORMS}
+        await _async_process_config(hass, conf)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_RELOAD, reload_service_handler, schema=vol.Schema({})
+    )
+
+    return await _async_process_config(hass, config)
+
+
+async def _async_process_config(hass, config) -> bool:
+    """Process rest configuration."""
+    if DOMAIN not in config:
+        return True
+
+    load_tasks = []
+    for conf in config[DOMAIN]:
+        scan_interval = conf.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+        resource_template = conf.get(CONF_RESOURCE_TEMPLATE)
+        rest = create_rest_data_from_config(hass, conf)
+        coordinator = _wrap_rest_in_coordinator(
+            hass, rest, resource_template, scan_interval
+        )
+        await coordinator.async_refresh()
+
+        for platform_domain in COORDINATOR_AWARE_PLATFORMS:
+            shared_data = hass.data[DOMAIN][platform_domain]
+            for shared_data_id, platform_conf in enumerate(
+                conf.get(platform_domain, [])
+            ):
+                shared_data[shared_data_id] = {REST: rest, COORDINATOR: coordinator}
+
+                load = discovery.async_load_platform(
+                    hass,
+                    platform_domain,
+                    DOMAIN,
+                    {SHARED_DATA_ID: shared_data_id, **platform_conf},
```

Comment:
I'm still not super happy with the way `platform_conf` is being passed to the platform. I haven't come up with a better way though.

Reply:
I could `enumerate` `config[DOMAIN]` as well and then pass this as `REST_IDX` and `PLATFORM_IDX`

Reply:
Looks like this definitely needs to change since if there is a template in the config it will cause an error with the recorder as the fired event won't be serializable

I'm going to store it in hass.data instead and pass a key to find it in the discovery info

Reply:
I reworked this.